src/parser/core/modules/NormalisedEvents.ts

```diff
@@ -0,0 +1,178 @@
+	type: 'normalisedheal'
+	targetsHit: number
+	hits: number
+	ghostedHits: number
+	ghostedAmount: number
+	calculatedEvents: HealEvent[]
+	confirmedEvents: HealEvent[]
+	attachEvent(event: HealEvent): void
+}
+
+export class NormalisedEvents extends Module {
+	static handle = 'normalisedEvents'
+
+	private static GenerateNewDamageEvent(event: DamageEvent): NormalisedDamageEvent {
+		return {
+			...event,
+			type: 'normaliseddamage',
+			targetsHit: 0,
+			hits: 0,
+			ghostedHits: 0,
+			ghostedAmount: 0,
+			calculatedEvents: [],
+			confirmedEvents: [],
+			attachEvent(event: DamageEvent): void {
+				if (event.type.includes('calculated')) {
+					this.calculatedEvents.push(event)
+				} else {
+					this.confirmedEvents.push(event)
+				}
+			},
+		}
+	}
+
+	private static GenerateNewHealEvent(event: HealEvent): NormalisedHealEvent {
+		return {
+			...event,
+			type: 'normalisedheal',
+			targetsHit: 0,
+			hits: 0,
+			ghostedHits: 0,
+			ghostedAmount: 0,
+			calculatedEvents: [],
+			confirmedEvents: [],
+			attachEvent(event: HealEvent): void {
+				if (event.type.includes('calculated')) {
+					this.calculatedEvents.push(event)
+				} else {
+					this.confirmedEvents.push(event)
+				}
+			},
+		}
+	}
+
+	@dependency private hitType!: HitType
+
+	_normalisedDamageEvents = new Map<number, NormalisedDamageEvent>()
+	_normalisedHealEvents = new Map<number, NormalisedHealEvent>()
+
+	normalise(events: Event[]): Event[] {
+		events.forEach((event) => {
+			if (DAMAGE_EVENT_TYPES.includes(event.type)) {
+				this.normaliseDamageEvent(event as DamageEvent)
+			}
+
+			if (HEAL_EVENT_TYPES.includes(event.type)) {
+				this.normaliseHealEvent(event as HealEvent)
+			}
+		})
+
+		this._normalisedDamageEvents.forEach(this.summarize, this)
+		this._normalisedHealEvents.forEach(this.summarize, this)
+
+		return events.concat(Array.from(this._normalisedDamageEvents.values()), Array.from(this._normalisedHealEvents.values()))
+	}
+
+	private normaliseDamageEvent(event: DamageEvent) {
+		let normalisedEvent = this.findRelatedDamageEvent(event)
+
+		if (!normalisedEvent) {
+			normalisedEvent = NormalisedEvents.GenerateNewDamageEvent(event)
+			this._normalisedDamageEvents.set(event.packetID ?? event.timestamp, normalisedEvent)
+		}
+
+		normalisedEvent.attachEvent(event)
+	}
+
+	private normaliseHealEvent(event: HealEvent) {
+		let normalisedEvent = this.findRelatedHealEvent(event)
+
+		if (!normalisedEvent) {
+			normalisedEvent = NormalisedEvents.GenerateNewHealEvent(event)
+			this._normalisedHealEvents.set(event.packetID ?? event.timestamp, normalisedEvent)
+		}
+
+		normalisedEvent.attachEvent(event)
+	}
+
+	private findRelatedDamageEvent(event: DamageEvent) {
+		if (event.packetID) {
+			return this._normalisedDamageEvents.get(event.packetID)
+		}
+
+		// No packetID set - match based on timestamps
+		const possibleRelatedEvents = Array.from(this._normalisedDamageEvents.values())
+			.filter((normalisedEvent) => {
```

Comment:
nits:

arrows w/ no type and one arg can forgo the parens (`.filter(normalisedEvent => ...)`).

---

you saw this one coming; can afford to remove the `{ return }` from this. It can still be split over multiple lines, just saves the fn body itself.